src/dataflow/src/render/delta_join.rs

```diff
@@ -232,70 +262,84 @@ where
+                                            Err(trace) => build_lookup(update_stream, trace.enter_region(region), prev_key, closure),
                                         }
                                     } else {
                                         match arrangements_neu.get(&(&inputs[*other], &next_key[..])).unwrap() {
-                                            Ok(local) => build_lookup(update_stream, local.enter_region(region), prev_key),
-                                            Err(trace) => build_lookup(update_stream, trace.enter_region(region), prev_key),
+                                            Ok(local) => build_lookup(update_stream, local.enter_region(region), prev_key, closure),
+                                            Err(trace) => build_lookup(update_stream, trace.enter_region(region), prev_key, closure),
                                         }
                                     };
                                     update_stream = oks;
                                     region_errs.push(errs);
 
-                                    // Update our map of the sources of each column in the update stream.
-                                    source_columns
-                                        .extend(input_mapper.global_columns(*other));
-
-                                    let (oks, errs) = build_filter(
-                                        update_stream,
-                                        &source_columns,
-                                        &mut predicates,
-                                        &mut equivalences,
-                                    );
-                                    update_stream = oks;
-                                    if let Some(errs) = errs {
-                                        region_errs.push(errs);
-                                    }
-
                                     bound_inputs.push(*other);
                                 }
 
-                                // We must now de-permute the results to return to the common order.
-                                // TODO: Non-demanded columns would need default values here.
-                                let permutation = (0 .. source_columns.len()).map(|c| {
-                                    source_columns.iter().position(|x| &c == x).expect("Did not find required column in output")
-                                }).collect::<Vec<_>>();
-                                update_stream = update_stream.map({
-                                    let mut row_packer = repr::RowPacker::new();
-                                    move |row| {
-                                        let datums = row.unpack();
-                                        row_packer.pack(permutation.iter().map(|c| datums[*c]))
-                                }});
+                                // We must now apply `mfp` as it *should* have sufficient support for
```

Comment:
Could this section somehow be collapsed with the very similar-looking section in render_join?

Reply:
Maybe. I'm going to rewrite it, as I now realize that you can have errors in literals, and so the `expect` calls may actually produce errors because maybe you just wanted to put a `1/0` in a column and idk but it could happen.

Reply:
I think superficially the code is similar, but as written at least the error term gets pushed in to a vec here vs concat'd previously. The following fragment is common to both, and maybe could get extracted but it is pretty small now (and the function signature might be as large as the body). I've written similar code a few times across a few PRs, and I'm up for extracting some commonality but might want to wait a bit to see how the rest of the methods shake out (e.g. there is something similar in `mfp_sources`).
```rust
let (updates, errors) = update_stream.flat_map_fallible({
    // Reuseable allocation for unpacking.
    let mut datums = DatumVec::new();
    let mut row_packer = repr::RowPacker::new();
    move |row| {
        let temp_storage = RowArena::new();
        let mut datums_local = datums.borrow_with(&row);
        // TODO(mcsherry): re-use `row` allocation.
        mfp.evaluate(&mut datums_local, &temp_storage, &mut row_packer)
            .map_err(DataflowError::from)
            .transpose()
    }
});
```